Exclude fast moving players when finding potential spawn points

ColinM9991 opened the pull request:
Resolves #155 

Introduces several new configuration values to impose vehicle, altitude and speed limits on players to ensure civilians aren't spawned around them if they're moving quickly, or too high.

The new filter function checks the height and speed of players and, if it exceeds the limit, they will be excluded from the shuffled players array that is subsequently used to retrieve a list of potential spawn points.

~~I feel the naming of `spawnLimitEnabled`, `spawnHeightLimit` and `spawnSpeedLimit` are contentious and may not be indicative of their true nature of limiting spawn points based on player speed, height and vehicle type.~~

ColinM9991 commented:
Another option to filtering, which I can do tomorrow if it's preferable, is to instead change the function to deal with a single player, then call it within the loop from `findSpawnPos`. Should reduce looping.

Fusselwurm commented:
> I feel the naming of `spawnLimitEnabled`, `spawnHeightLimit` and `spawnSpeedLimit` are contentious and may not be indicative of their true nature of limiting spawn points based on player speed, height and vehicle type.

Yes, we have a conflict between the wish for short variable names… and the complicated functionality behind them :grimacing: 

My first approach would be to center the naming around us *ignoring* players when looking for spawn points.

Maybe `spawnDisregardsSomePlayers` / `spawnDisregardsPlayersAboveHeight` / `spawnDisregardsPlayersAboveSpeed` ?

Or `dontSpawnForSomePlayers` / `dontSpawnForPlayersHigherThan` / `dontSpawnForPlayersFasterThan` ?

Fusselwurm commented on addons/lifecycle/functions/fnc_filterPlayers.sqf:
```diff
@@ -0,0 +1,35 @@
+
+LOG_1("%1 is enabled, continuing", QGVAR(spawnLimitEnabled));
+
+private _filteredPlayers = [];
+
+{
+    private _playerVehicle = vehicle _x;
+
+    // If player is not in a vehicle, or is not in an air vehicle
+    if (_playerVehicle isEqualTo _x || {!(_playerVehicle isKindOf "Air")}) then {
+        _filteredPlayers pushBackUnique _x;
+
+        continue;
+    };
+
+    private _playerHeight = getPosATL _playerVehicle # 2;
+    private _playerSpeed = speed _playerVehicle;
+
+    private _isSuitableCandidate = _playerHeight <= GVAR(spawnHeightLimit) && {_playerSpeed <= GVAR(spawnSpeedLimit)};
+
+    LOG_2("Player %1 candidate status is %2", _x, _isSuitableCandidate);
+    
+    if (_isSuitableCandidate) then { _filteredPlayers pushBackUnique _x };
+} forEach _allPlayers;
+
+_filteredPlayers
```
Using `forEach` & add to a new array within the loop certainly works.

I think it would be more expressive - and a bit less code - if we were to use `select` instead 

```suggestion
_allPlayers select {
    private _playerVehicle = vehicle _x;

    // If player is not in a vehicle, or is not in an air vehicle
    if (_playerVehicle isEqualTo _x || {!(_playerVehicle isKindOf "Air")}) then { 
        true
    } else {
        private _playerHeight = getPosATL _playerVehicle # 2;
        private _playerSpeed = speed _playerVehicle; 
        _playerHeight <= GVAR(spawnHeightLimit) && {_playerSpeed <= GVAR(spawnSpeedLimit)}
    }
}
```

ColinM9991 replied:
Refactored to use `select`. Also changed `vehicle player` to `objectParent` as it's slightly quicker and more reliable when dealing with corpses.

Fusselwurm commented:
you know what… we dont even need to restrict ourselves to filtering out airborne players.

If I were to drive a car at 200km/h, I should probably also get ignored.

Fusselwurm reviewed (approved):
I like it.

I'll let this stew for another day or three before merging, maybe someone comes up with better variable names.

ColinM9991 commented:
> I like it.
> 
> I'll let this stew for another day or three before merging, maybe someone comes up with better variable names.

I've removed the checks for the type of vehicle. Might possibly be worth benchmarking in the future to check against CfgVehicles like mortars etc, but it may be unnecessary performance tuning for now as it shouldn't really be that detrimental.

Renamed the variables to `spawnCandidateLimitEnabled` and `spawnCandidateXXXLimit` as they're quite short and get the point across.

Fusselwurm commented:
> Might possibly be worth benchmarking in the future to check against CfgVehicles like mortars etc, but it may be unnecessary performance tuning for now as it shouldn't really be that detrimental.

the spawn cycle defaults to running once every two seconds, I think it will do fine :D
